kotsu/run.py

```diff
@@ -0,0 +1,60 @@
+"""Interface for running a registry of models on a registry of validations."""
+from typing import Tuple
+
+import logging
+import time
+
+from kotsu import store
+from kotsu.registration import ModelRegistry, ValidationRegistry
+
+
+logger = logging.getLogger(__name__)
+
+
+def run(
+    model_registry: ModelRegistry,
+    validation_registry: ValidationRegistry,
+    store_directory: str,
+    run_params: dict = {},
+):
+    """Run a registry of models on a registry of validations.
+
+    Args:
+        model_registry: A ModelRegistry containing the registry of models to be run through
+            validations.
+        validation_registry: A ValidationRegistry containing the registry of validations to run
+            each model through.
+        store_directory: A file path or URI location to store the validation results and any extra
+            output artefacts of the validations and models.
+        run_params: A dictionary of optional run parameters.
+    """
+    results_list = []
+    for validation_spec in validation_registry.all():
+        for model_spec in model_registry.all():
+            logger.info(f"Running validation - model: {validation_spec.id} - {model_spec.id}")
+            validation = validation_spec.make()
+            model = model_spec.make()
+            results, elapsed_secs = _run_validation_model(validation, model, run_params)
+            results.update(
+                {
+                    "validation_id": validation_spec.id,
+                    "model_id": model_spec.id,
+                    "runtime_secs": elapsed_secs,
+                }
+            )
```

Comment:
I think that there could be some improvements in the typing of results. In my mind, it is not clear as to what a validation should be returning and how this interacts with the result of other validations.

It might make things clearer if there was another object called ResultsList that took the responsibility for:
- Holding the results of the validation
- Holding the metadata about the run of that validation
- Converting this data into a data frame

I could imagine a default `ResultsList` class with an interface:
```
class ResultsList():
    def __init__(results_cols):
    # The results cols would then define the expected properties of the validations results data
    ...

    def validation_result(run_time, dict_with_results_cols, validation_spec, model_spec)
    
    def to_dataframe():
     # This would be replacement of the code in store that builds the results dataframe
```

A `ResultsList` would then be passed to the `run` function. The object would validate the output of the validations. The object, with the results, would then be the returned object from the `run` function.

Reply:
re the validation return type, we could either check the results at run time in some way (eg as suggested above), or we could take a leaf out of eg Pydantic's book and do some runtime checking of the type of the function (at registration time) - eg, _demand_ that validation functions are typed, and maybe even demand they return a namedtuple or whatever so we can even know the columns ahead of time

Reply:
Thoughts added to #19